libsplinter/src/admin/store/yaml/mod.rs

```diff
@@ -0,0 +1,1830 @@
+            // write circuit
+            store.write_circuit_state()?;
+            // read proposals
+            store.read_proposal_state()?;
+        } else {
+            // write all empty state
+            store.write_state()?;
+        }
+
+        Ok(store)
+    }
+
+    /// Read circuit state from the circuit file path and cache the contents in the store
+    fn read_circuit_state(&mut self) -> Result<(), YamlAdminStoreError> {
+        let circuit_file = File::open(&self.circuit_file_path).map_err(|err| {
+            YamlAdminStoreError::general_error_with_source(
+                "Failed to open YAML circuit state file",
+                Box::new(err),
+            )
+        })?;
+
+        let yaml_state_circuits: YamlCircuitState = serde_yaml::from_reader(&circuit_file)
+            .map_err(|err| {
+                YamlAdminStoreError::general_error_with_source(
+                    "Failed to read YAML circuit state file",
+                    Box::new(err),
+                )
+            })?;
+
+        let yaml_state = CircuitState::from(yaml_state_circuits);
+
+        let mut state = self.state.lock().map_err(|_| {
+            YamlAdminStoreError::general_error("YAML admin service store's internal lock poisoned")
+        })?;
+
+        for (circuit_id, circuit) in yaml_state.circuits.iter() {
+            for service in circuit.roster.iter() {
+                let service_id =
+                    ServiceId::new(service.service_id.to_string(), circuit_id.to_string());
+
+                state.service_directory.insert(service_id, service.clone());
+            }
+        }
+
+        state.circuit_state = yaml_state;
+        Ok(())
+    }
+
+    /// Read circuit proposal state from the proposal file path and cache the contents in the
+    /// store
+    fn read_proposal_state(&mut self) -> Result<(), YamlAdminStoreError> {
+        let proposal_file = File::open(&self.proposal_file_path).map_err(|err| {
+            YamlAdminStoreError::general_error_with_source(
+                "Failed to open YAML proposal state file",
+                Box::new(err),
+            )
+        })?;
+
+        let proposals_state: ProposalState =
+            serde_yaml::from_reader(&proposal_file).map_err(|err| {
+                YamlAdminStoreError::general_error_with_source(
+                    "Failed to read YAML proposal state file",
+                    Box::new(err),
+                )
+            })?;
+
+        let mut state = self.state.lock().map_err(|_| {
+            YamlAdminStoreError::general_error("YAML admin service store's internal lock poisoned")
+        })?;
+
+        state.proposal_state = proposals_state;
+        Ok(())
+    }
+
+    /// Read circuit state from the circuit file path and cache the contents in the store and then
+    /// read circuit proposal state from the proposal file path and cache the contents in the
+    /// store
+    fn read_state(&mut self) -> Result<(), YamlAdminStoreError> {
```

Comment:
Can this method just call `read_circuit_state` and `read_proposal_state` instead of duplicating them? Same for the write method.

Reply:
I had seperated them out like this so the state lock would be held during the required read/write. If we just called read_circuit_state/read_proposal_state the lock could be acquired inbetween.